### DISCUSSION TITLE
Is there a maximum upload size for inserts via the serverless driver?

### DISCUSSION THREAD

**Discussion by PhilBookst** [planetscale/discussion#384] [Q&A] [answered]
Is there a maximum upload size for uploads via the serverless driver?

I cannot find any info in the docs - are there any limits?

Thank you!

**Comment by orware** [comment]
Hello @PhilBookst,

I haven't run similar tests using the serverless driver, but a few months back I ran some tests related to storing a large amount of JSON within a column and the results there indicated that you could `INSERT` about 16 MiB of JSON at a time, so I would expect the maximum upload size via the serverless driver to be somewhat similar to that.

For our databases, running `SHOW VARIABLES LIKE 'max_allowed_packet';` results in a value of 64 MiB being returned currently, but that may serve a different purpose as I'll share next.

The main drawback of storing so much data within a single row is that it could impact your querying, unless you are always retrieving one row at a time, since in my testing that day, retrieving more than a few rows at these larger sizes resulted in the `max_allowed_packet` value being exceeded.

This could result in your database being overwhelmed more easily and lead to errors like the following:

```
SQL Error [1153] [42000]: target: test-database.-.primary: vttablet: rpc error: code = ResourceExhausted desc = grpc: trying to send message larger than max (94357261 vs. 67108864)
SQL Error [1105] [HY000]: target: test-database.-.primary: vttablet: rpc error: code = Unavailable desc = error reading from server: EOF
```

If you do run some extended tests using the serverless driver I'll be curious what your results are, but hopefully the limits are similar to what I've shared above since I would expect them to be essentially the same as when connecting via a typical MySQL client at the moment.

**Comment by mattrobenolt** [comment, marked as answer]
It's slightly tricky to quantify this, but at the HTTP level, our limit is 100MB for the entire request. But this may be compressed with gzip or something allowing you to technically send more than 100MB. But as @orware mentions, there are limits against the underlying mysqlds and other internal systems that may hit limits that are a bit harder to predict purely based on your HTTP request.

Since it's not 1:1 on how your HTTP request translates into how it hits mysqld or how it's passed along to vitess, it's really hard to predict exactly what a limit would be in practice.

I'd say in general though, probably not good practice to try to be sending 50MB in a single database query. :)
